daml-script/runner/src/main/scala/com/digitalasset/daml/lf/engine/script/Runner.scala

```diff
@@ -53,6 +57,82 @@ object LfValueCodec extends ApiCodecCompressed[AbsoluteContractId](false, false)
+    party_participants: Map[Party, Participant],
+) {
+  def getPartyParticipant(party: Party): Either[String, T] =
+    party_participants.get(party) match {
+      case None =>
+        default_participant.fold[Either[String, T]](
+          Left(s"No participant for party $party and no default participant"))(Right(_))
+      case Some(participant) => getParticipant(Some(participant))
+    }
+  def getParticipant(participantOpt: Option[Participant]): Either[String, T] =
+    participantOpt match {
+      case None =>
+        default_participant.fold[Either[String, T]](Left(s"No default participant"))(Right(_))
+      case Some(participant) =>
+        participants.get(participant) match {
+          case None =>
+            default_participant.fold[Either[String, T]](
+              Left(s"No participant $participant and no default participant"))(Right(_))
+          case Some(t) => Right(t)
+        }
+    }
+}
+
+object ParticipantsJsonProtocol extends DefaultJsonProtocol {
+  implicit object ParticipantFormat extends JsonFormat[Participant] {
+    def read(value: JsValue) = value match {
+      case JsString(s) => Participant(s)
+      case _ => deserializationError("Expected Participant string")
+    }
+    def write(p: Participant) = JsString(p.participant)
+  }
+  implicit object PartyFormat extends JsonFormat[Party] {
+    def read(value: JsValue) = value match {
+      case JsString(s) => Party(s)
+      case _ => deserializationError("Expected Party string")
+    }
+    def write(p: Party) = JsString(p.party)
+  }
+  implicit val apiParametersFormat = jsonFormat2(ApiParameters)
+  implicit val participantsFormat = jsonFormat3(Participants[ApiParameters])
+}
+
+object Runner {
+  private def connectApiParameters(params: ApiParameters, clientConfig: LedgerClientConfiguration)(
+      implicit ec: ExecutionContext,
+      seq: ExecutionSequencerFactory): Future[LedgerClient] = {
+    LedgerClient.singleHost(params.host, params.port, clientConfig)
+  }
+  // We might want to have one config per participant at some point but for now this should be sufficient.
+  def connect(
+      participantParams: Participants[ApiParameters],
+      clientConfig: LedgerClientConfiguration)(
+      implicit ec: ExecutionContext,
+      seq: ExecutionSequencerFactory): Future[Participants[LedgerClient]] = {
+    for {
+      // The standard library is incredibly weird. Option is not Traversable so we have to convert to a list and back.
+      // Map is but it doesn’t return a Map so we have to call toMap afterwards.
+      defaultClient <- Future
+        .traverse(participantParams.default_participant.toList)(x =>
+          connectApiParameters(x, clientConfig))
+        .map(_.headOption)
+      participantClients <- Future
+        .traverse(participantParams.participants: Map[Participant, ApiParameters])({
+          case (k, v) => connectApiParameters(v, clientConfig).map((k, _))
+        })
+        .map(_.toMap)
```

Comment:
Occurs because traverse's type parameters are wrong, as expanded a little in #6469: https://github.com/digital-asset/daml/pull/6469/files#diff-911de6a24e211785457135a0d167f519R173-R178